gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientReadChannel.java

```diff
@@ -0,0 +1,568 @@
+   * which contradicts {@link SeekableByteChannel#position(long) SeekableByteChannel} contract.
+   * TODO(user): decide if this needs to be fixed.
+   *
+   * @param newPosition the new position, counting the number of bytes from the beginning.
+   * @return this channel instance
+   * @throws FileNotFoundException if the underlying object does not exist.
+   * @throws IOException on IO error
+   */
+  @Override
+  public SeekableByteChannel position(long newPosition) throws IOException {
+    throwIfNotOpen();
+
+    if (newPosition == currentPosition) {
+      return this;
+    }
+
+    validatePosition(newPosition);
+    logger.atFiner().log(
+        "Seek from %s to %s position for '%s'", currentPosition, newPosition, resourceId);
+    currentPosition = newPosition;
+    return this;
+  }
+
+  @Override
+  public long size() throws IOException {
+    return objectSize;
+  }
+
+  @Override
+  public SeekableByteChannel truncate(long size) throws IOException {
+    throw new UnsupportedOperationException("Cannot mutate read-only channel");
+  }
+
+  @Override
+  public boolean isOpen() {
+    return open;
+  }
+
+  @Override
+  public void close() throws IOException {
+    try {
+      if (open) {
+        logger.atFiner().log("Closing channel for '%s'", resourceId);
+        contentReadChannel.closeContentChannel();
+      }
+    } catch (Exception e) {
+      throw new IOException(
+          String.format("Exception occurred while closing channel '%s'", resourceId), e);
+    } finally {
+      contentReadChannel = null;
+      open = false;
+    }
+  }
+
+  /**
+   * This class own the responsibility of opening up contentChannel. It also implements the Fadvise,
+   * which helps in deciding the boundaries of content channel being opened and also caching the
+   * footer of an object.
+   */
+  private class ContentReadChannel {
+
+    // Size of buffer to allocate for skipping bytes in-place when performing in-place seeks.
+    private static final int SKIP_BUFFER_SIZE = 8192;
+    private final BlobId blobId;
+
+    // This is the actual current position in `contentChannel` from where read can happen.
+    // This remains unchanged of position(long) method call.
+    private long contentChannelCurrentPosition = -1;
+    private long contentChannelEnd = -1;
+    // Prefetched footer content.
+    private byte[] footerContent;
+    // Used as scratch space when reading bytes just to discard them when trying to perform small
+    // in-place seeks.
+    private byte[] skipBuffer = null;
+    private ReadableByteChannel byteChannel = null;
+    private boolean randomAccess;
+
+    public ContentReadChannel(
+        GoogleCloudStorageReadOptions readOptions, StorageResourceId resourceId) {
+      this.blobId =
+          BlobId.of(
+              resourceId.getBucketName(), resourceId.getObjectName(), resourceId.getGenerationId());
+      this.randomAccess = readOptions.getFadvise().equals(Fadvise.RANDOM);
```

Comment:
You do not need `equals` method for enums, it's idiomatic to use `==`:
```java
this.randomAccess = readOptions.getFadvise() == Fadvise.RANDOM;
```

Reply:
Updated.